File: bench/limit_crossval.zig
const bcs = @import("bcs");
const std = @import("std");

const D0 = struct { leaf: u8 };
const D1 = struct { child: D0 };
const D2 = struct { child: D1 };
const D3 = struct { child: D2 };
const D4 = struct { child: D3 };
const D5 = struct { child: D4 };
const D6 = struct { child: D5 };
const D7 = struct { child: D6 };
const D8 = struct { child: D7 };

fn deepValue() D8 {
    return .{
        .child = .{
            .child = .{
                .child = .{
                    .child = .{
                        .child = .{
                            .child = .{
                                .child = .{
                                    .child = .{ .leaf = 7 },
                                },
                            },
                        },
                    },
                },
            },
        },
    };
}

fn errName(err: bcs.Error) []const u8 {
    return switch (err) {
        error.ContainerTooDeep => "container_too_deep",
        error.NotSupported => "not_supported",
        else => "other",
    };
}

fn emitBytesResult(name: []const u8, result: bcs.Error![]u8, allocator: std.mem.Allocator) void {
    if (result) |bytes| {
        defer allocator.free(bytes);
        std.debug.print("{s}=ok:", .{name});
        for (bytes) |byte| std.debug.print("{x:0>2}", .{byte});
        std.debug.print("\n", .{});
    } else |err| {
        std.debug.print("{s}=err:{s}\n", .{ name, errName(err) });
    }
}

fn emitSizeResult(name: []const u8, result: bcs.Error!usize) void {
    if (result) |size| {
        std.debug.print("{s}=ok:{}\n", .{ name, size });
    } else |err| {
        std.debug.print("{s}=err:{s}\n", .{ name, errName(err) });
    }
}

fn emitUnitResult(name: []const u8, result: anytype) void {
    if (result) |_| {
        std.debug.print("{s}=ok\n", .{name});
    } else |err| {
        std.debug.print("{s}=err:{s}\n", .{ name, errName(err) });
    }
}

pub fn main() !void {
    const allocator = std.heap.c_allocator;
    const value = deepValue();
    const limit_fail: usize = 8;
    const limit_ok: usize = 9;
    const bytes = try bcs.serializeWithLimit(allocator, value, limit_ok);
    defer allocator.free(bytes);

    emitBytesResult("to_bytes_with_limit_fail", bcs.serializeWithLimit(allocator, value, limit_fail), allocator);
    emitBytesResult("to_bytes_with_limit_ok", allocator.dupe(u8, bytes), allocator);
    emitBytesResult(
        "to_bytes_with_limit_above_max",
        bcs.serializeWithLimit(allocator, value, @as(usize, bcs.max_container_depth) + 1),
        allocator,
    );

    emitSizeResult("serialized_size_with_limit_fail", bcs.serializedSizeWithLimit(value, limit_fail));
    emitSizeResult("serialized_size_with_limit_ok", bcs.serializedSizeWithLimit(value, limit_ok));
    emitSizeResult(
        "serialized_size_with_limit_above_max",
        bcs.serializedSizeWithLimit(value, @as(usize, bcs.max_container_depth) + 1),
    );

    emitUnitResult("from_bytes_with_limit_fail", bcs.deserializeWithLimit(D8, allocator, bytes, limit_fail));
    emitUnitResult("from_bytes_with_limit_ok", bcs.deserializeWithLimit(D8, allocator, bytes, limit_ok));
    emitUnitResult(
        "from_bytes_with_limit_above_max",
        bcs.deserializeWithLimit(D8, allocator, bytes, @as(usize, bcs.max_container_depth) + 1),
    );
}

File: bench/limit_crossval.rs
use bcs::{
    from_bytes_with_limit, serialized_size_with_limit, to_bytes_with_limit, Error,
    MAX_CONTAINER_DEPTH,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D0 {
    leaf: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D1 {
    child: D0,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D2 {
    child: D1,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D3 {
    child: D2,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D4 {
    child: D3,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D5 {
    child: D4,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D6 {
    child: D5,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D7 {
    child: D6,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct D8 {
    child: D7,
}

fn deep_value() -> D8 {
    D8 {
        child: D7 {
            child: D6 {
                child: D5 {
                    child: D4 {
                        child: D3 {
                            child: D2 {
                                child: D1 {
                                    child: D0 { leaf: 7 },
                                },
                            },
                        },
                    },
                },
            },
        },
    }
}

fn hex(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|b| format!("{:02x}", b))
        .collect::<Vec<_>>()
        .join("")
}

fn err_name(err: &Error) -> &'static str {
    match err {
        Error::ExceededContainerDepthLimit(_) => "container_too_deep",
        Error::NotSupported(_) => "not_supported",
        _ => "other",
    }
}

fn emit_bytes_result(name: &str, result: Result<Vec<u8>, Error>) {
    match result {
        Ok(bytes) => println!("{name}=ok:{}", hex(&bytes)),
        Err(err) => println!("{name}=err:{}", err_name(&err)),
    }
}

fn emit_size_result(name: &str, result: Result<usize, Error>) {
    match result {
        Ok(size) => println!("{name}=ok:{size}"),
        Err(err) => println!("{name}=err:{}", err_name(&err)),
    }
}

fn emit_unit_result<T>(name: &str, result: Result<T, Error>) {
    match result {
        Ok(_) => println!("{name}=ok"),
        Err(err) => println!("{name}=err:{}", err_name(&err)),
    }
}

fn main() {
    const LIMIT_FAIL: usize = 8;
    const LIMIT_OK: usize = 9;

    let value = deep_value();
    let bytes = to_bytes_with_limit(&value, LIMIT_OK).unwrap();

    emit_bytes_result(
        "to_bytes_with_limit_fail",
        to_bytes_with_limit(&value, LIMIT_FAIL),
    );
    emit_bytes_result("to_bytes_with_limit_ok", Ok(bytes.clone()));
    emit_bytes_result(
        "to_bytes_with_limit_above_max",
        to_bytes_with_limit(&value, MAX_CONTAINER_DEPTH + 1),
    );

    emit_size_result(
        "serialized_size_with_limit_fail",
        serialized_size_with_limit(&value, LIMIT_FAIL),
    );
    emit_size_result(
        "serialized_size_with_limit_ok",
        serialized_size_with_limit(&value, LIMIT_OK),
    );
    emit_size_result(
        "serialized_size_with_limit_above_max",
        serialized_size_with_limit(&value, MAX_CONTAINER_DEPTH + 1),
    );

    emit_unit_result(
        "from_bytes_with_limit_fail",
        from_bytes_with_limit::<D8>(&bytes, LIMIT_FAIL),
    );
    emit_unit_result(
        "from_bytes_with_limit_ok",
        from_bytes_with_limit::<D8>(&bytes, LIMIT_OK),
    );
    emit_unit_result(
        "from_bytes_with_limit_above_max",
        from_bytes_with_limit::<D8>(&bytes, MAX_CONTAINER_DEPTH + 1),
    );
}
